### DISCUSSION TITLE
It's time to start plantin'! Parametric History Tree with Searchbar for entities!

### DISCUSSION THREAD

**Discussion by Arch-druid** [hlorus/CAD_Sketcher#181] [Feature Requests]
**Describe & Illustrate your Feature Request**
FreeCAD and Autodesk Fusion 360 both have a very intuitive building tree system in which bodies/singular solids are maintained within a tree. The developers and I would really have to sit down to discuss what would constitute the architecture of the tree, although having a form of organization to reference subcomponents of a solid body would be very powerful.
**Motivation / Existing Problem**
The problem with not having a tree primarily exists in the navigation through very complex models. All lines and constraint points are simply left on the selection menu. Even if the raw python code variable were given, it would at least allow for some degree of recognition as to which face/edge/vertex belongs to what

Description
The first thing we need to do is make a parent-child system in the tree, and the first thing that could use a parent child system is the sketching operation. Basically, when a new sketch is created, all lines should go under the "Sketch" parent. Right now, the entities box is something that will get progressively disorganized and harder to navigate with complicated designs. All entities belonging to a sketch, for example, should be maintained within that sketch. All sketches that are part of solid bodies should be maintained within a solid body parent. However, all children must ALSO be able to be copied and pasted OUTSIDE of their parents. For example, a cube body would have a child sketch that is simply a square, and that sketch would have a child entity that is a line. Any line in the sketch should be capable of being copied and pasted into another sketch or body, or as a standalone component. 

Alternative Solutions
Not really an alternative here. The tree requires a UI change and a bit of nested functionality

Mockup![image](https://user-images.githubusercontent.com/65218891/173421643-a80fde9d-0279-45ae-9314-a8c267648cd5.png)

**Comment by rljacobson** [comment]
I am not sure I fully understand this feature proposal. In my mind, there are two distinct but _potentially_ related features: 

1. Structuring models into sub-models ("pieces," "parts," "components") organized hierarchically, where items specific to a sub-model like sketches can be grouped together.
2. Recording the operations used to construct a model. This would potentially allow:
		- previous operations to be modified and the construction to be replayed, propagating forward the modification so it is reflected in subsequent model states
		- deletion of intermediate operations and propagating the change forward (a version of the previous)
		- a course-grained form of "undo" (same as previous, but applying only to the last operation performed)
		- "branching" of the history of the construction to be able to compare different construction results

I view feature 1 to be essential and feature 2 to be _extremely valuable_. There are infinitely many variations on how something like feature 2 could work. FreeCAD sort of combines both ideas into a single UI. Fusion360 has them as distinct UI concepts. Treating it conceptually as a "history" versus an operation stack, deciding which user actions constitute a node in the history/operation stack, the UI representation… in my view these are [bike shed colors](https://en.wikipedia.org/wiki/Law_of_triviality).

I am just a hobbyist and am very new to Blender altogether, coming from Fusion360 and (to a much lesser extent) FreeCAD. I know nothing. So take my comments with the gravity they deserve. 😀

**Comment by jaherbert1** [comment]
Adding another vote for @rljacobson's 1.  I've noticed that are are few similar requests on the discord as well (search: "sketch groups").  This is essential for even moderately large projects.

Custom groups for sketches would also allow the creation of a collapsed group to deposit (and hide) the (possibly many) outdated sketches that cannot be deleted for fear of invoking the dreaded ["Delete dimensional constraint with drivers"](https://github.com/hlorus/CAD_Sketcher/issues/318) bug.

I should also mention that maintaining the flat, chronological sketch list view as an alternative would be desirable.
